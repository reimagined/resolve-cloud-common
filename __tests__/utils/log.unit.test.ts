import {
  getLog,
  setLogCorrelationId,
  resetLogCorrelationId,
  logRootNamespace
} from '../../utils/index'

const prettify = ({ mock }: any): any =>
  mock.calls
    .map((str: string) => `${str}`.trim())
    .join('\n')
    .replace(/\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z)/gi, '')
    .replace(/\+\d+ms/gi, '')
    // eslint-disable-next-line no-control-regex
    .replace(/[^m]+m/g, '')
    .replace(/\n /g, '\n')
    .replace(/^ /g, '')
    .replace(/ {2}/g, '')

const writeTestLogs = (log: any): any => {
  const largeObject = {
    a: {
      b: {
        c: {
          arr: [
            1,
            2,
            3,
            4,
            5,
            6,
            7,
            8,
            {
              a: {
                b: {
                  c: true
                }
              }
            }
          ]
        }
      }
    }
  }

  log.log('test text')
  log.log(42)
  log.log(largeObject)

  log.error('test text')
  log.error(42)
  log.error(largeObject)

  log.warn('test text')
  log.warn(42)
  log.warn(largeObject)

  log.debug('test text')
  log.debug(42)
  log.debug(largeObject)

  log.info('test text')
  log.info(42)
  log.info(largeObject)

  log.verbose('test text')
  log.verbose(42)
  log.verbose(largeObject)
}

for (const { describeName, prepare } of [
  {
    describeName: 'log',
    prepare: (): void => {
      resetLogCorrelationId()
    }
  },
  {
    describeName: 'log with correlationId',
    prepare: (): void => {
      setLogCorrelationId('test-correlation-id')
    }
  }
]) {
  describe(describeName, () => {
    let originalStandardWrite: any
    let originalErrorWrite: any
    // eslint-disable-next-line func-names
    const write = jest.fn().mockImplementation(function (...args) {
      originalStandardWrite.call(this, ...args)
    })
    beforeAll(() => {
      originalStandardWrite = process.stdout.write
      process.stdout.write = write.bind(process.stdout)
      originalErrorWrite = process.stderr.write
      process.stderr.write = write.bind(process.stderr)
    })

    afterAll(() => {
      process.stdout.write = originalStandardWrite
      process.stderr.write = originalErrorWrite
    })

    beforeEach(() => {
      prepare()
    })

    afterEach(() => {
      delete process.env.DEBUG_LEVEL
      write.mockClear()
    })

    test(`should not write to scope "test-scope"`, () => {
      const scope = 'test-scope'
      const log = getLog(scope)
      writeTestLogs(log)

      expect(prettify(write)).toMatchSnapshot()
    })

    test(`should write to scope "${logRootNamespace}:my-test" (log,error,warn,debug,info)`, () => {
      const log = getLog(`${logRootNamespace}:my-test`)
      writeTestLogs(log)

      expect(prettify(write)).toMatchSnapshot()
    })

    test(`should write to scope "${logRootNamespace}:my-test" (log,error)`, () => {
      process.env.DEBUG_LEVEL = 'error'

      const log = getLog(`${logRootNamespace}:my-test`)
      writeTestLogs(log)

      expect(prettify(write)).toMatchSnapshot()
    })

    test(`should write to scope "resolve:*" (log,error,warn,debug,info,verbose) (resolve:*)`, () => {
      process.env.DEBUG = 'resolve:*'
      process.env.DEBUG_LEVEL = 'verbose'

      const logStorage = getLog(`resolve:storage`)
      const logBus = getLog(`resolve:bus`)
      const logCustomer = getLog(`customer`)

      for (const log of [logStorage, logBus, logCustomer]) {
        writeTestLogs(log)
      }

      expect(prettify(write)).toMatchSnapshot()
    })
  })
}
