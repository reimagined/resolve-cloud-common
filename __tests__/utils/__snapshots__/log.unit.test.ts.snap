// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`log should not write to scope "test-scope" 1`] = `""`;

exports[`log should write to scope "resolve:*" (log,error,warn,debug,info,verbose) (resolve:*) 1`] = `
"resolve:storage LOG <no-correlation-id> test text 
resolve:storage LOG <no-correlation-id> 42 
resolve:storage LOG <no-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:storage ERROR <no-correlation-id> test text 
resolve:storage ERROR <no-correlation-id> 42 
resolve:storage ERROR <no-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:storage WARN <no-correlation-id> test text 
resolve:storage WARN <no-correlation-id> 42 
resolve:storage WARN <no-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:storage DEBUG <no-correlation-id> test text 
resolve:storage DEBUG <no-correlation-id> 42 
resolve:storage DEBUG <no-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:storage INFO <no-correlation-id> test text 
resolve:storage INFO <no-correlation-id> 42 
resolve:storage INFO <no-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:storage VERBOSE <no-correlation-id> test text 
resolve:storage VERBOSE <no-correlation-id> 42 
resolve:storage VERBOSE <no-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:bus LOG <no-correlation-id> test text 
resolve:bus LOG <no-correlation-id> 42 
resolve:bus LOG <no-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:bus ERROR <no-correlation-id> test text 
resolve:bus ERROR <no-correlation-id> 42 
resolve:bus ERROR <no-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:bus WARN <no-correlation-id> test text 
resolve:bus WARN <no-correlation-id> 42 
resolve:bus WARN <no-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:bus DEBUG <no-correlation-id> test text 
resolve:bus DEBUG <no-correlation-id> 42 
resolve:bus DEBUG <no-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:bus INFO <no-correlation-id> test text 
resolve:bus INFO <no-correlation-id> 42 
resolve:bus INFO <no-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:bus VERBOSE <no-correlation-id> test text 
resolve:bus VERBOSE <no-correlation-id> 42 
resolve:bus VERBOSE <no-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} "
`;

exports[`log should write to scope "resolve:cloud:lambda:environment:my-test" (log,error) 1`] = `
"resolve:cloud:lambda:environment:my-test LOG <no-correlation-id> test text 
resolve:cloud:lambda:environment:my-test LOG <no-correlation-id> 42 
resolve:cloud:lambda:environment:my-test LOG <no-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:cloud:lambda:environment:my-test ERROR <no-correlation-id> test text 
resolve:cloud:lambda:environment:my-test ERROR <no-correlation-id> 42 
resolve:cloud:lambda:environment:my-test ERROR <no-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} "
`;

exports[`log should write to scope "resolve:cloud:lambda:environment:my-test" (log,error,warn,debug,info) 1`] = `
"resolve:cloud:lambda:environment:my-test LOG <no-correlation-id> test text 
resolve:cloud:lambda:environment:my-test LOG <no-correlation-id> 42 
resolve:cloud:lambda:environment:my-test LOG <no-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:cloud:lambda:environment:my-test ERROR <no-correlation-id> test text 
resolve:cloud:lambda:environment:my-test ERROR <no-correlation-id> 42 
resolve:cloud:lambda:environment:my-test ERROR <no-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:cloud:lambda:environment:my-test WARN <no-correlation-id> test text 
resolve:cloud:lambda:environment:my-test WARN <no-correlation-id> 42 
resolve:cloud:lambda:environment:my-test WARN <no-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:cloud:lambda:environment:my-test DEBUG <no-correlation-id> test text 
resolve:cloud:lambda:environment:my-test DEBUG <no-correlation-id> 42 
resolve:cloud:lambda:environment:my-test DEBUG <no-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:cloud:lambda:environment:my-test INFO <no-correlation-id> test text 
resolve:cloud:lambda:environment:my-test INFO <no-correlation-id> 42 
resolve:cloud:lambda:environment:my-test INFO <no-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} "
`;

exports[`log with correlationId should not write to scope "test-scope" 1`] = `""`;

exports[`log with correlationId should write to scope "resolve:*" (log,error,warn,debug,info,verbose) (resolve:*) 1`] = `
"resolve:storage LOG <test-correlation-id> test text 
resolve:storage LOG <test-correlation-id> 42 
resolve:storage LOG <test-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:storage ERROR <test-correlation-id> test text 
resolve:storage ERROR <test-correlation-id> 42 
resolve:storage ERROR <test-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:storage WARN <test-correlation-id> test text 
resolve:storage WARN <test-correlation-id> 42 
resolve:storage WARN <test-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:storage DEBUG <test-correlation-id> test text 
resolve:storage DEBUG <test-correlation-id> 42 
resolve:storage DEBUG <test-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:storage INFO <test-correlation-id> test text 
resolve:storage INFO <test-correlation-id> 42 
resolve:storage INFO <test-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:storage VERBOSE <test-correlation-id> test text 
resolve:storage VERBOSE <test-correlation-id> 42 
resolve:storage VERBOSE <test-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:bus LOG <test-correlation-id> test text 
resolve:bus LOG <test-correlation-id> 42 
resolve:bus LOG <test-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:bus ERROR <test-correlation-id> test text 
resolve:bus ERROR <test-correlation-id> 42 
resolve:bus ERROR <test-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:bus WARN <test-correlation-id> test text 
resolve:bus WARN <test-correlation-id> 42 
resolve:bus WARN <test-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:bus DEBUG <test-correlation-id> test text 
resolve:bus DEBUG <test-correlation-id> 42 
resolve:bus DEBUG <test-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:bus INFO <test-correlation-id> test text 
resolve:bus INFO <test-correlation-id> 42 
resolve:bus INFO <test-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:bus VERBOSE <test-correlation-id> test text 
resolve:bus VERBOSE <test-correlation-id> 42 
resolve:bus VERBOSE <test-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} "
`;

exports[`log with correlationId should write to scope "resolve:cloud:lambda:environment:my-test" (log,error) 1`] = `
"resolve:cloud:lambda:environment:my-test LOG <test-correlation-id> test text 
resolve:cloud:lambda:environment:my-test LOG <test-correlation-id> 42 
resolve:cloud:lambda:environment:my-test LOG <test-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:cloud:lambda:environment:my-test ERROR <test-correlation-id> test text 
resolve:cloud:lambda:environment:my-test ERROR <test-correlation-id> 42 
resolve:cloud:lambda:environment:my-test ERROR <test-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} "
`;

exports[`log with correlationId should write to scope "resolve:cloud:lambda:environment:my-test" (log,error,warn,debug,info) 1`] = `
"resolve:cloud:lambda:environment:my-test LOG <test-correlation-id> test text 
resolve:cloud:lambda:environment:my-test LOG <test-correlation-id> 42 
resolve:cloud:lambda:environment:my-test LOG <test-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:cloud:lambda:environment:my-test ERROR <test-correlation-id> test text 
resolve:cloud:lambda:environment:my-test ERROR <test-correlation-id> 42 
resolve:cloud:lambda:environment:my-test ERROR <test-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:cloud:lambda:environment:my-test WARN <test-correlation-id> test text 
resolve:cloud:lambda:environment:my-test WARN <test-correlation-id> 42 
resolve:cloud:lambda:environment:my-test WARN <test-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:cloud:lambda:environment:my-test DEBUG <test-correlation-id> test text 
resolve:cloud:lambda:environment:my-test DEBUG <test-correlation-id> 42 
resolve:cloud:lambda:environment:my-test DEBUG <test-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} 
resolve:cloud:lambda:environment:my-test INFO <test-correlation-id> test text 
resolve:cloud:lambda:environment:my-test INFO <test-correlation-id> 42 
resolve:cloud:lambda:environment:my-test INFO <test-correlation-id> {\\"a\\":{\\"b\\":{\\"c\\":{\\"arr\\":[1,2,3,4,5,6,7,8,{\\"a\\":{\\"b\\":{\\"c\\":true}}}]}}}} "
`;
